src/main/java/dev/morling/onebrc/CalculateAverage_mariuslarsen.java

```diff
@@ -0,0 +1,160 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package dev.morling.onebrc;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class CalculateAverage_mariuslarsen {
+    private static final Byte DELIMITER = ';';
+    private static final Byte NEWLINE = '\n';
+    private static final int N_THREADS = Runtime.getRuntime().availableProcessors();
+    private static final int MAX_NUMBER_OF_DESTINATIONS = 10000;
+    private static final int MIN_BLOCK_SIZE = 1 << 15;
+    private static final int MAX_LINE_WIDTH = 108;
+
+    public static void main(String[] args) throws IOException {
+        try (FileChannel channel = FileChannel.open(Path.of("./measurements.txt"), StandardOpenOption.READ)) {
+            Map<String, Stats> res = createBuffers(channel).parallelStream()
+                    .map(CalculateAverage_mariuslarsen::parseBuffer)
+                    .flatMap(Collection::stream)
+                    .collect(Collectors.toMap(s -> new String(s.city), Function.identity(), Stats::join, TreeMap::new));
+            System.out.println(res);
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static List<MappedByteBuffer> createBuffers(FileChannel channel) throws IOException {
+        long channelSize = channel.size();
+        long blockSize;
+        if (N_THREADS * MIN_BLOCK_SIZE > channelSize) {
+            blockSize = MIN_BLOCK_SIZE;
+        }
+        else {
+            long wantedBlockSize = Math.ceilDiv(channelSize, N_THREADS) + MAX_LINE_WIDTH;
+            blockSize = Math.min(Integer.MAX_VALUE, wantedBlockSize);
+        }
+        List<MappedByteBuffer> blocks = new ArrayList<>();
+        long pos = 0;
+        while (pos < channelSize) {
+            int size = (int) Math.min(blockSize, channelSize - pos);
+            MappedByteBuffer b = channel.map(FileChannel.MapMode.READ_ONLY, pos, size);
+            while (b.get(--size) != '\n')
+                ;
+            b.limit(size + 1);
+            blocks.add(b);
+            pos += size + 1;
+        }
+        return blocks;
+    }
+
+    private static Collection<Stats> parseBuffer(ByteBuffer buffer) {
+        Map<Integer, Stats> destinations = HashMap.newHashMap(MAX_NUMBER_OF_DESTINATIONS);
+        Stats currentStats;
+        int size;
+        int hashCode;
+        int negative;
+        int temperature;
+        byte current;
+        int offset;
+        while (buffer.hasRemaining()) {
+            hashCode = 1;
+            offset = buffer.position();
+            while ((current = buffer.get()) != DELIMITER) {
+                hashCode = 31 * hashCode + current;
```

Comment:
How would that deal with hash collisions?

Reply:
Badly 😅 
That may explain why the solution was pretty fast. I'll create a new PR when I find a better solution.